### PR TITLE
c8d: Fix getting the platform of a single-platform image

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

https://github.com/moby/moby/pull/46495 changed the code to look for the platform when the image is a single-platform one but it looked at the wrong place, we now look at the image config  to get the platform when the image is not multi-platform

**- How I did it**

**- How to verify it**

This test should still work, also `docker --runtime io.containerd.wasmedge.v1 --platform wasi/wasm rumpl/bb` should run just fine too

```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestCreateDifferentPlatform TEST_INTEGRATION_USE_SNAPSHOTTER=1 TEST_IGNORE_CGROUP_CHECK=1 test-integration
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
This only errors if an explicit `--platform` is provided (and it doesn't match), correct?

ISTR for the non-containerd variant we produced a _warning_ if no explicit platform was passed for these (but I'd have to go and check); wondering if we need something similar here.

(although it's possible that BuildKit already broke that scenario)

File: daemon/containerd/image.go [hunk 1]
```diff
@@ -157,17 +157,26 @@ func (i *ImageService) GetImageManifest(ctx context.Context, refOrID string, opt
 
 	cs := i.client.ContentStore()
 
-	desc, err := i.resolveDescriptor(ctx, refOrID)
+	image, err := i.resolveImage(ctx, refOrID)
```
Comment (thaJeztah) [inline daemon/containerd/image.go:160]: This variable collides with the `image` import

File: daemon/containerd/image.go [hunk 2]
```diff
@@ -157,17 +157,26 @@ func (i *ImageService) GetImageManifest(ctx context.Context, refOrID string, opt
...
 	if err != nil {
 		return nil, err
 	}
 
+	desc := image.Target
 	if containerdimages.IsManifestType(desc.MediaType) {
+		plat := desc.Platform
+		if plat == nil {
+			config, err := image.Config(ctx, cs, platform)
```
Comment (thaJeztah) [inline daemon/containerd/image.go:169]: For a follow-up, I wonder if  some of this needs to be combined with `resolveDescriptor` and/or `resolveImage`.  I had a very quick glance, and it looks like in some cases we already use those, and directly after do some platform matching; so perhaps having a `platform` argument that can be passed could be useful.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left one nit for the variable name (yes, still need to look at enabling that linter 😂)

**Comment by thaJeztah** [comment]
Docker hub looks to be a bit sad today;

```
ERROR: failed to solve: distribution/distribution:2.8.2: failed to authorize: failed to fetch oauth token: unexpected status from POST request to https://auth.docker.io/token: 503 Service Unavailable
Error: buildx bake failed with: ERROR: failed to solve: distribution/distribution:2.8.2: failed to authorize: failed to fetch oauth token: unexpected status from POST request to https://auth.docker.io/token: 503 Service Unavailable
```

File: daemon/containerd/image.go (hunk 2 above)
Comment (rumpl) [reply daemon/containerd/image.go:169]: This is actually wrong, once we get the descriptor of the image config we need to `readConfig` to get the real config

File: daemon/containerd/image.go (hunk 1 above)
Comment (rumpl) [reply daemon/containerd/image.go:160]: Changed it thanks, yeah it would be great if we were able to enable this lint

**Comment by rumpl** [comment]
> ISTR for the non-containerd variant we produced a _warning_ if no explicit platform was passed for these (but I'd have to go and check); wondering if we need something similar here.

The non-containerd variant produces a warning only if a specific platform was given 

https://github.com/moby/moby/blob/b94e88c1e29f929facdb6fe5b18a4d9416c268be/daemon/images/image.go#L203-L205

File: daemon/containerd/image.go (hunk 2 above)
Comment (rumpl) [reply daemon/containerd/image.go]: We definitely need a nicer API to get the bits we usually need for sure

File: daemon/containerd/image.go [hunk 3]
```diff
@@ -157,17 +157,32 @@ func (i *ImageService) GetImageManifest(ctx context.Context, refOrID string, opt
...
 	if containerdimages.IsManifestType(desc.MediaType) {
+		plat := desc.Platform
+		if plat == nil {
+			config, err := img.Config(ctx, cs, platform)
+			if err != nil {
+				return nil, err
+			}
+			var configPlatform ocispec.Platform
+			if err := readConfig(ctx, cs, config, &configPlatform); err != nil {
+				return nil, err
```
Comment (vvoland) [inline daemon/containerd/image.go:175]: Should we have `cerrdefs.IsNotFound` here more and provide a better errors?
Currently that will result in `content digest X not found` error when requesting platform not present locally:

```bash
$ docker run --pull never --platform linux/s390x --rm -it busybox
docker: Error response from daemon: content digest sha256:7d602b12b1d9c1bdbf4c9255c0ba276ac0d9e0cd781a7c13461e4875cfcae509: not found.
See 'docker run --help'.
```

**Review by vvoland — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**